### PR TITLE
feat(type-safe-api): websocket apis for typescript

### PR DESCRIPTION
**Pull request by cogwirrel**
Add a new `TypeSafeWebSocketApiProject` which can be used to build WebSocket APIs in a similar manner to REST APIs.

This includes:

* Modelling an API in Smithy or OpenAPI
* Generated CDK infrastructure (WebSocket API and Lambda functions)
* Generated runtimes containing lambda handler wrappers, and server SDK to send messages to clients
* Generated handler stubs with the `@handler` trait
* Generated client and react hooks libraries for use client side
* Integration with infrastructure and website projects
* WebSocket API explorer
* WebSocket API CDK construct with validation schemas as custom resource to scale for larger apis

Only TypeScript is supported currently, but refactoring work has been done to make Python and Java support more straightforward.

Note the new Java project in `internal/smithy-async-transformer` - this allows us to perform a transform on users' Smithy models to augment operations with additional traits required for OpenAPI generation.

Note that this upgrades the version of CDK we depend on so that we can use the WebSocketApi construct - this has had a few minor knock on effects on the pipeline package.

Remaining items for follow up PRs:
* Documentation generation from model
* Python support
* Java support

### CODE REVIEW

File: packages/cloudscape-react-ts-website/samples/cloudscape-react-ts-website/src/components/TypeSafeWebSocketApiClientProvider/index.tsx.mustache [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  {{/typeSafeWebSocketApis}}
+
+  return {{#typeSafeWebSocketApis}}{{{apiNameSafe}}}Client{{^isLast}} && {{/isLast}}{{/typeSafeWebSocketApis}} ? (
+    {{#typeSafeWebSocketApis}}
+    {{=<% %>=}}
+    <<%apiNameSafe%>WebSocketApiClientProvider client={<%apiNameSafe%>Client}>
+    <%={{ }}=%>
+    {{/typeSafeWebSocketApis}}
+      {children}
+    {{#typeSafeWebSocketApisReversed}}
```
Comment (agdimech) [inline packages/cloudscape-react-ts-website/samples/cloudscape-react-ts-website/src/components/TypeSafeWebSocketApiClientProvider/index.tsx.mustache:25]: This code looks somewhat familiar :)

File: packages/cloudscape-react-ts-website/samples/cloudscape-react-ts-website/src/components/WebSocketApiExplorer/index.tsx.mustache [hunk 2]
```diff
@@ -0,0 +1,738 @@
+import { Sha256 } from "@aws-crypto/sha256-js";
```
Comment (agdimech) [inline packages/cloudscape-react-ts-website/samples/cloudscape-react-ts-website/src/components/WebSocketApiExplorer/index.tsx.mustache:1]: OOf - this is quite a hefty file. DO you foresee changes to this in future/worth extracting into a lib of some sort?

File: packages/cloudscape-react-ts-website/src/cloudscape-react-ts-website-project.ts [hunk 3]
```diff
@@ -129,10 +155,24 @@ export class CloudscapeReactTsWebsiteProject extends ReactTypeScriptProject {
...
+        ?.replace(/[^a-z0-9_]+/gi, "")
+        .replace(/^[0-9]+/gi, ""),
+      hooksPackage:
+        tsApi.library?.typescriptWebSocketHooks?.package?.packageName,
+      clientPackage:
+        tsApi.library?.typescriptWebSocketClient?.package?.packageName,
+    }));
     const mustacheConfig = {
       applicationName: this.applicationName,
+      hasApis: apis.length + webSocketApis.length > 0,
```
Comment (agdimech) [inline packages/cloudscape-react-ts-website/src/cloudscape-react-ts-website-project.ts:171]: TILI: rename apis to typeSafeApis

File: packages/infrastructure/samples/infrastructure/typescript/src/constructs/apis/websocketapi.ts.mustache [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    // Grant authenticated users access to interact with the WebSocket API
+    props?.userIdentity.identityPool.authenticatedRole.addToPrincipalPolicy(
+      new PolicyStatement({
+        effect: Effect.ALLOW,
+        actions: ["execute-api:Invoke"],
+        resources: [
+          Stack.of(this).formatArn({
+            service: "execute-api",
+            resource: this.api.api.apiId,
+            resourceName: "*",
```
Comment (agdimech) [inline packages/infrastructure/samples/infrastructure/typescript/src/constructs/apis/websocketapi.ts.mustache:47]: is this OK from a nagpack perspective?

File: packages/type-safe-api/docs/developer_guides/type-safe-api/websocket_clients.md [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+
+=== "TS"
+
+    ```ts
+    import { DefaultApiWebSocketClient } from "myapi-typescript-websocket-client";
+
+    const client = await DefaultApiWebSocketClient.connect(...);
+
+    // Send the message to the server. The method is the same as the operation name defined in your model.
+    await client.sayHello({ name: "Jack" });
```
Comment (agdimech) [inline packages/type-safe-api/docs/developer_guides/type-safe-api/websocket_clients.md:87]: leaving your mark ;)

File: packages/type-safe-api/docs/developer_guides/type-safe-api/websocket_clients.md [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+
+=== "TS"
+
+    ```ts
+    import { DefaultApiWebSocketClient } from "myapi-typescript-websocket-client";
+
+    const client = await DefaultApiWebSocketClient.connect(...);
+
+    // Listen to greetings. A function is returned which can be used to remove the listener.
+    const unlistenToGreetings = await client.onSendGreeting((input) => {
```
Comment (agdimech) [inline packages/type-safe-api/docs/developer_guides/type-safe-api/websocket_clients.md:116]: This is super clean - nice job!

File: packages/type-safe-api/docs/developer_guides/type-safe-api/websocket_lambda_handlers.md [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+      // `connectionId` is the ID of the new connection
+      // `sdk` is used to send messages to connected clients
+      // Note that you cannot send messages to the new connection until after this function returns
+      const { connectionId, sdk } = request;
+
+      // TODO: Implement
+
+      // Use the below to allow or deny the incoming connection (when the lambda returns).
+      // The connection is allowed by default.
+      if (Math.random() > 0.5) {
```
Comment (agdimech) [inline packages/type-safe-api/docs/developer_guides/type-safe-api/websocket_lambda_handlers.md]: maybe replace this with something like: `if (!authenticated(request)) ...`

File: packages/type-safe-api/docs/developer_guides/type-safe-api/websocket_using_openapi.md [hunk 8]
```diff
@@ -0,0 +1,99 @@
+# Using OpenAPI
+
+As an alternative to [Smithy](./using_smithy.md), you can use [OpenAPI Version 3.0.3](https://swagger.io/specification/) to define your API.
+
+!!!warning
+
+    Note that while Type Safe WebSocket API uses the OpenAPI format to define operations, many of the features of OpenAPI are not supported by WebSockets and therefore may be ignored if you define them in your model, for example specifying `parameters` or `responses`.
```
Comment (agdimech) [inline packages/type-safe-api/docs/developer_guides/type-safe-api/websocket_using_openapi.md:7]: We should probably call out explicitly that Smithy is the recommended model lang.

**Review by agdimech — APPROVED** [review]
What an achievement!

This is an epic new feature and love how clean and concise the constructs and docs are. Well done!

File: packages/cloudscape-react-ts-website/samples/cloudscape-react-ts-website/src/components/WebSocketApiExplorer/index.tsx.mustache (hunk 2 above)
Comment (cogwirrel) [reply packages/cloudscape-react-ts-website/samples/cloudscape-react-ts-website/src/components/WebSocketApiExplorer/index.tsx.mustache:1]: I think it'd definitely be worth extracting to a lib in the future! Not sure if NorthStar would be the best home... I wonder whether we look at adding non jsii package publishing to PDK in the future?

File: packages/cloudscape-react-ts-website/src/cloudscape-react-ts-website-project.ts (hunk 3 above)
Comment (cogwirrel) [reply packages/cloudscape-react-ts-website/src/cloudscape-react-ts-website-project.ts:171]: Might leave it - we already have `this.typeSafeApis` which are the project instances, and then `apis` and `webSocketApis` are the mustache configs for each :)

File: packages/type-safe-api/docs/developer_guides/type-safe-api/websocket_lambda_handlers.md (hunk 7 above)
Comment (cogwirrel) [reply packages/type-safe-api/docs/developer_guides/type-safe-api/websocket_lambda_handlers.md]: Good shout!

File: packages/infrastructure/samples/infrastructure/typescript/src/constructs/apis/websocketapi.ts.mustache (hunk 4 above)
Comment (cogwirrel) [reply packages/infrastructure/samples/infrastructure/typescript/src/constructs/apis/websocketapi.ts.mustache:47]: Ooh there are a few nag warnings actually - will address these!

**Pull request merged**